src/maintenance.js

```diff
@@ -0,0 +1,159 @@
+      {expires: newNamespace.expires, contact: newNamespace.contact},
+      {expires, contact})) {
+      await newNamespace.modify(entity => {
+        entity.expires = expires;
+        entity.contact = contact;
+      });
+
+      newNamespace = await Namespace.load({namespace: namespace});
+    }
+  }
+
+  if (created) {
+    // set up the first user as active,
+    await setPulseUser({
+      username: `${namespace}-1`,
+      password: newNamespace.password,
+      namespace, cfg, rabbitManager});
+    // ..and the second user as inactive (empty string means no logins allowed)
+    await setPulseUser({
+      username: `${namespace}-2`,
+      password: '',
+      namespace, cfg, rabbitManager});
+  }
+
+  return newNamespace;
+};
+
+module.exports.delete = async function({Namespace, rabbitManager, cfg, namespace}) {
+  let debug = Debug('delete');
+
+  // use the configuration regexp to determine if an object is owned by this user
+  let owned = new RegExp(cfg.app.userConfigPermission.replace(/{{namespace}}/g, namespace));
+
+  // find the user's exchanges and queues
+  let exchanges = _.filter(await rabbitManager.exchanges(),
+    e => e.vhost === cfg.app.virtualhost && owned.test(e.name));
+  let queues = _.filter(await rabbitManager.queues(),
+    q => q.vhost === cfg.app.virtualhost && owned.test(q.name));
+
+  // delete sequentually to avoid overloading the rabbitmq server
+  for (let i = 0; i < exchanges.length; i++) {
+    let name = exchanges[i].name;
+    debug(`deleting exchange ${name}`);
+    await rabbitManager.deleteExchange(name, cfg.app.virtualhost);
+  }
+  for (let i = 0; i < queues.length; i++) {
+    let name = queues[i].name;
+    debug(`deleting queue ${name}`);
+    await rabbitManager.deleteQueue(name, cfg.app.virtualhost);
+  }
+
+  // try to delete both users
+  await rabbitManager.deleteUser(`${namespace}-1`, cfg.app.virtualhost);
+  await rabbitManager.deleteUser(`${namespace}-2`, cfg.app.virtualhost);
+
+  // finally, delete the table row
+  Namespace.remove({namespace});
+};
+
+module.exports.expire = async function({Namespace, cfg, rabbitManager, now}) {
+  let count = 0;
+  let debug = Debug('expire');
+
+  await Namespace.scan({
+    expires: Namespace.op.lessThan(now),
+  }, {
+    limit:            250, // max number of concurrent delete operations
+    handler:          async (ns) => {
+      count++;
+      debug(`deleting expired namespace ${ns.namespace}`);
+      await module.exports.delete({Namespace, rabbitManager, cfg, namespace: ns.namespace});
+    },
+  });
+  return count;
+};
+
+module.exports.rotate = async function({Namespace, now, cfg, rabbitManager}) {
+  let count = 0;
+  let debug = Debug('rotate');
+
+  await Namespace.scan({
+    nextRotation: Namespace.op.lessThan(now),
+  }, {
+    limit:            250, // max number of concurrent modify operations
+    handler:          async (ns) => {
+      count++;
+      debug(`rotating ${ns.namespace}`);
+      let password = slugid.v4();
```

Comment:
Now that I think about it, we don't make any crypto guarantees about slugids in the slugid readme, afaict. Is it a safe generator to use for this purpose? I'm 99.9% sure it is fine given the use case, but perhaps we should make a note of our reasoning somewhere.